src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -1973,17 +1970,36 @@ public void run()
                     if (logger.isTraceEnabled())
                         logger.trace("Async instance tidier for {}, after barrier", descriptor);
 
-                    if (bf != null)
-                        bf.close();
-                    if (summary != null)
-                        summary.close();
-                    if (runOnClose != null)
+                    Throwable exceptions = null;
+                    if (runOnClose != null) try
+                    {
                         runOnClose.run();
-                    if (dfile != null)
-                        dfile.close();
-                    if (ifile != null)
-                        ifile.close();
-                    globalRef.release();
+                    }
+                    catch (RuntimeException | Error ex)
+                    {
+                        logger.error("Failed to run on-close listeners for sstable " + descriptor.baseFilename(), ex);
+                        exceptions = ex;
+                    }
+
+                    Throwable closeExceptions = Throwables.close(null, closeables);
```

Comment:
Isn't `FileUtils.close` a better match for what is done here?

Reply:
I'm not sure TBH, need to get the exceptions reported and log them, add them to the `exceptions`. With `FileUtils.close` I'd have to catch the exception.

Reply:
It's practically the same pattern, since you do have a check for this concrete part throwing exceptions.

The reason for asking is to avoid relaxing `Throwables.close` -- we may lose an error message we care about for something that should not be null.

Alternatively, we can add a `closeIgnoringNulls` method to `Throwables`.

Reply:
9ffff01f36d271686bac16b932b57c7d918a355d

Reply:
looks good